### ISSUE TITLE
[2019-12][interpreter] Crash running tests

### ISSUE THREAD

**Issue by spouliot** [mono/mono#18273]
http://xamarin-storage/jenkins/xamarin-macios/pull-request/a0c657a850b5eefbe5709d2b6fe6804009dc4f0f/3334590/device-tests/jenkins-results/tests/index.html

## Steps to Reproduce

1. run `monotouch-test` with the interpreter

### Current Behavior

The 3 configurations using the interpreter is crashing

### Expected Behavior

No crash

## On which platforms did you notice this

[x] iOS
[ ] macOS
[ ] Linux
[ ] Windows

**Version Used**:

2019-12
This is not seen in 2019-10 (or earlier)

## Stacktrace

http://xamarin-storage/jenkins/xamarin-macios/pull-request/a0c657a850b5eefbe5709d2b6fe6804009dc4f0f/3334590/device-tests/jenkins-results/tests/monotouch-test/665/monotouchtest-2019-12-19-104153.symbolicated.log

**Comment by spouliot** [comment]
c.c. @SamMonoRT (2019-12) and @lewurm (interpreter)

**Comment by spouliot** [comment]
The same 3 configurations also fails for tvOS (makes sense)
http://xamarin-storage/jenkins/xamarin-macios/pull-request/a0c657a850b5eefbe5709d2b6fe6804009dc4f0f/3334592/device-tests/jenkins-results/tests/index.html

**Comment by lewurm** [comment]
It's asserting here:

https://github.com/mono/mono/blob/07fee8a91dbc172bacded9d2c7f7d158eb7c50f1/mono/utils/mono-threads-posix.c#L326-L330

which was just introduced recently by the non-recursive interpreter change in https://github.com/mono/mono/pull/16461 (cherry-picked from https://github.com/dotnet/coreclr/pull/1057 ).


`status` is `-1` and `errno` is set to `[EINVAL]` where the `man` page of `mprotect` has to say the following:
> `addr` is not a multiple of the page size (i.e.  `addr` is not page-aligned).

Well, that's not true:
```
(lldb) p/x memory_barrier_process_wide_helper_page
(void *) $0 = 0x0000000108229000
```
and `PAGESIZE = 0x1000`.

It seems like `mprotect` just doesn't work on iOS.

-----------------------------------------------------
So I traced back what happened in the linux kernel with this workaround. When changing the permissions of a page it does this:
https://github.com/torvalds/linux/blob/596cf45cbf6e4fa7bcb0df33e373a7d062b644b5/mm/mprotect.c#L289-L291
```c
	/* Only flush the TLB if we actually modified any entries: */
	if (pages)
		flush_tlb_range(vma, start, end);
```
where 
https://github.com/torvalds/linux/blob/cef7298262e9af841fb70d8673af45caf55300a1/arch/arm64/include/asm/tlbflush.h#L88-L93
```c
/*	flush_tlb_range(vma, start, end)
 *		Invalidate the virtual-address range '[start, end)' on all
 *		CPUs for the user address space corresponding to 'vma->mm'.
 *		Note that this operation also invalidates any walk-cache
 *		entries associated with translations for the specified address
 *		range.                         */
```
where its implementation eventually issues a sequence of `dsb ishst; tlbi <mode>; dsb ish`. The `tlbi` instruction does the TLB invalidation which we want to achieve here, however, we can't execute this instruction with user space privileges.

So we need to find a way how to trick the iOS kernel into doing the same via some syscalls.

**Comment by lewurm** [comment]
***TL;DR***: Things should be fine after the pagesize thing is fixed.

------------------------------------

> and `PAGESIZE = 0x1000`.
> 
> It seems like `mprotect` just doesn't work on iOS.

That's wrong. Alas `PAGESIZE` is just hardcoded in that file (in case for iOS), instead of using `mono_pagesize ()` which determines the system's pagesize properly. The actual pagesize is `0x4000`, thus the address from my previous comment `0x0000000108229000` is not page-aligned. That's _real_ reason why `mprotect` fails.

I figured that out by trying different things, but eventually I was confused why the current solution _sometimes_ succeeds executing the memory barrier but then ends up in an access violation: Assume the helper page is actually page aligned, then `mprotect` succeeds. However, the remaining `0x3000` bytes are used by `malloc` (`posix_memalign ()` is provided by the malloc implementation), but `mprotect` will change the permission of the _whole_ page.

--------------------------------------------------------------------------

From what I gathered when reading through the XNU code, similar things (`dsb` and `tlbi` instructions) should be executed via `mprotect` as on Linux. `mprotect` will eventually end up in `pmap_protect` which is arch specific: https://github.com/apple/darwin-xnu/blob/a449c6a3b8014d9406c2ddbdc81795da24aa7443/osfmk/arm/pmap.c#L5550-L5551
```c
		FLUSH_PTE_RANGE_STRONG(bpte_p, epte_p);
		PMAP_UPDATE_TLBS(pmap, start, end);
```
which is defined as
https://github.com/apple/darwin-xnu/blob/a449c6a3b8014d9406c2ddbdc81795da24aa7443/osfmk/arm/pmap.c#L809-L812
```c
#define PMAP_UPDATE_TLBS(pmap, s, e) {							\
	flush_mmu_tlb_region_asid_async(s, (unsigned)(e - s), pmap);			\
	sync_tlb_flush();								\
}
```
-> 
https://github.com/apple/darwin-xnu/blob/a449c6a3b8014d9406c2ddbdc81795da24aa7443/osfmk/arm/pmap.c#L9128-L9134
```c
flush_mmu_tlb_region_asid_async (...) {
        // [...]
	va = tlbi_asid(asid) | tlbi_addr(va);
	end = tlbi_asid(asid) | tlbi_addr(end);
	if (pmap->nested == TRUE) {
		flush_mmu_tlb_allentries_async(va, end);
	} else {
		flush_mmu_tlb_entries_async(va, end);
	}
        // [...]
}
```
->
https://github.com/apple/darwin-xnu/blob/a449c6a3b8014d9406c2ddbdc81795da24aa7443/osfmk/arm64/machine_routines_asm.s#L377-L387
```asm
/*
 *	void flush_mmu_tlb_entries_async(uint64_t, uint64_t)
 *
 *		Flush TLB entries, don't wait for completion
 */
	.text
	.align 2
	.globl EXT(flush_mmu_tlb_entries_async)
LEXT(flush_mmu_tlb_entries_async)
	FLUSH_MMU_TLB_ENTRIES
	ret
```
[...]
https://github.com/apple/darwin-xnu/blob/a449c6a3b8014d9406c2ddbdc81795da24aa7443/osfmk/arm64/machine_routines_asm.s#L356-L357
```asm
.macro FLUSH_MMU_TLB_ENTRIES
// [...]
1: // Lflush_mmu_tlb_entries_loop
	tlbi    vae1is, x0
// [...]
```
together with `sync_tlb_flush ()`:
https://github.com/apple/darwin-xnu/blob/a449c6a3b8014d9406c2ddbdc81795da24aa7443/osfmk/arm64/machine_routines_asm.s#L148-L164
```asm
.macro SYNC_TLB_FLUSH
	dsb	ish
	isb	sy
.endmacro

/*
 *	void sync_tlb_flush(void)
 *
 *		Synchronize one or more prior TLB flush operations
 */
	.text
	.align 2
	.globl EXT(sync_tlb_flush)
LEXT(sync_tlb_flush)
	SYNC_TLB_FLUSH
	ret
```
thus my userland guts conclude it should do the job we indent here.
